Add Archives block

talldan opened the pull request:
_This PR is a continuation of #5495, moved to a local branch for easier collaboration. Props @miina for the bulk of the implementation._

Uses #5602 

Continued archives block implementation from #1518 and #5495. Archive block implementation that uses similar output and filters as the archives widget. Has alignment toolbar within the block (`left`, `center`, `right`).

## Screenshots:
Block toolbar:
![screen shot 2018-05-01 at 2 23 23 pm](https://user-images.githubusercontent.com/3294597/39471330-4609a114-4d4b-11e8-8363-e9d48b8a6b1f.png)
Inspector toolbar:
![screen shot 2018-05-01 at 2 17 54 pm](https://user-images.githubusercontent.com/3294597/39471199-7c2b1148-4d4a-11e8-8629-478bf701b31d.png)
Dropdown view:
![archives-block-dropdown](https://user-images.githubusercontent.com/3294597/37142771-b714ed28-22c1-11e8-97a7-412d65e42df6.png)
List view:
![archives-block-list](https://user-images.githubusercontent.com/3294597/37142773-b74f6b2e-22c1-11e8-8dbb-229b533ff220.png)


Fixes #1464.

talldan committed:
Changes from PR #5495 squashed into single commit

Co-authored-by: Pascal Birchler <hello@pascalbirchler.ch>
Co-authored-by: Miina Sikk <miina.sikk@gmail.com>
Co-authored-by: Weston Ruter <weston@xwp.co>

Soean commented on core-blocks/archives/block.js:
```diff
@@ -0,0 +1,84 @@
+		this.toggleDisplayAsDropdown = this.toggleDisplayAsDropdown.bind( this );
+	}
+
+	toggleShowPostCounts() {
```
Can we use just one toggle function for both attributes? See: https://github.com/WordPress/gutenberg/pull/7941/files#diff-4a7082059a72e724bea52a0a00c65773R35

Soean commented on core-blocks/archives/index.php:
```diff
@@ -0,0 +1,105 @@
+	}
+
+	$block_content = sprintf(
+		'<div class="%1$s">%2$s</div>',
```
`div` should be an 'ul'

talldan replied:
Nice catch! It should be a `ul` when rendered as a list, and I've kept it as a div when rendered as a dropdown - 65025c1

talldan committed:
Remove archived keywords, since the title is implicitly a keyword already

talldan committed:
Use already destructured property `attributes` over `this.props.attributes`

talldan committed:
Use title case for copy

talldan committed:
Use `Disabled` element to prevent interaction with Archives Block editor view

Use of `pointer-events` to disable interaction was problematic. It
didn't disabled the dropdown version of the block and it doesn't prevent
keyboard interaction.

The Disabled component handles these situations.

talldan committed:
Rename block.js to edit.js for consistency with other blocks

talldan committed:
Convert ArchivesEdit to a function component

talldan committed:
Remove `isSelected` boolean and inline InspectorControls

talldan committed:
Render archives list container as `ul` when a list and `div` when a dropdown

talldan commented:
Hey @noisysocks & @mcsf - I think I've addressed all the review feedback from #5495 now. Would be great if one of you could look over may changes to make sure they're ok. Thanks!

mcsf committed:
Remove unnecessary React keys

mcsf committed:
Improve description copy

mcsf commented on core-blocks/archives/index.js:
```diff
@@ -0,0 +1,39 @@
+export const settings = {
+	title: __( 'Archives' ),
+
+	description: __( 'Display a monthly archive of your site’s Posts.' ),
```
FWIW, filter `widget_archives_dropdown_args` can change the periodic interval, meaning "monthly" may become inaccurate. However, I could see filtering as a more advanced feature, and sticking with "monthly" offers the benefit of conveying the basic idea of the block more clearly than e.g. "Display a periodic archive […]". Because of this, I don't mind sticking with the current copy.

mcsf reviewed (approved):
Tests well, looks good. Thanks for working on this, @talldan.
